
# Proyecto AEMET 
### Resumen de Extracción, Transformación y Carga (ETL) de Datos Climatológicos
---
Este notebook consolida y explica paso a paso todo el flujo de trabajo desarrollado para interactuar con la API de la AEMET (Agencia Estatal de Meteorología), procesar y limpiar la información, y exportar un dataset histórico listo para análisis o entrenamiento de modelos.

---
## 📋 Objetivos de este Notebook
1. **Autenticación e Integración:** Conectar con la API de AEMET OpenData utilizando una API Key almacenada de forma segura en un archivo `.env` local.
2. **Descarga del Catálogo de Estaciones:** Obtener y mostrar el maestro de estaciones meteorológicas operativas en España.
3. **Consulta de Datos de Prueba:** Descargar un subconjunto de datos diarios para una estación específica en un rango de fechas reducido para validar la conexión.
4. **Limpieza y Modelado de Datos (ETL):** Procesar la respuesta JSON de la API, reemplazando caracteres incorrectos, convirtiendo tipos de datos (strings a floats con coma, datetime) y manejando nulos de forma robusta.
5. **Descarga Masiva e Histórica:** Implementar un bucle para extraer datos históricos de los últimos años, gestionando de forma proactiva los límites de velocidad (rate limiting) de la API mediante retardos estratégicos.
6. **Exportación de Datos:** Consolidar el dataset completo en archivos Excel y CSV en el espacio local del usuario.
7. **Visualización Resumen:** Crear un gráfico de control que muestre la evolución del clima a lo largo del tiempo para validar los datos visualmente.
8. **Creacion de tablas:** Crear tablas para Postgree.
9. **Conexion y carga maestro a Postgree:** Conexion y carga de los datos de las estaciones como maestro.
10. **Conexion y carga datos estaciones entre fechas a Postgree:** Conexion y carga de los datos de las estaciones entre fechas.



---
### 📋 Imports necesarios para el Proyecto AEMET 📋 ###
---
### 📁 Sistema Operativo, Rutas y Archivos
* **`os`**: Gestionar directorios, rutas locales y variables del sistema.
* **`pathlib.Path`**: Manejar rutas de archivos de forma robusta e independiente de la plataforma (Windows / Linux).
* **`zipfile`**: Descomprimir y procesar archivos meteorológicos comprimidos en formato `.zip`.
* **`sys`**: Controlar parámetros específicos del intérprete de Python y del sistema.
* **`subprocess`**: Ejecutar herramientas, comandos o scripts externos directamente desde el flujo de ejecución.
* **`gc`**: Liberar memoria explícitamente y gestionar el recolector de basura (*garbage collector*) para optimizar recursos en procesos pesados.
* **`pickle`**: Serializar y deserializar objetos de Python para guardarlos y cargarlos desde archivos locales de forma rápida.

### ⏱️ Control de Tiempo, Fechas y Predicciones
* **`time` / `sleep`**: Añadir pausas controladas en la ejecución para evitar bloqueos por exceso de peticiones (*rate-limiting*).
* **`random`**: Generar números pseudoaleatorios (ideal para añadir variabilidad o *jitter* a los tiempos de espera entre peticiones a la API).
* **`datetime` / `timedelta`**: Manipular fechas para las predicciones e históricos meteorológicos de AEMET.
* **`calendar`**: Operar con meses y días para la agregación de estadísticas climáticas.

### 📊 Procesamiento y Análisis de Datos Climáticos
* **`pandas` (como `pd`)**: Cargar, limpiar y analizar series temporales, registros climáticos y tablas de datos.
* **`numpy` (como `np`)**: Realizar cálculos estadísticos vectorizados y operaciones matemáticas avanzadas.
* **`json`**: Procesar y formatear las respuestas estructuradas obtenidas desde la API.

### 🌐 Conexión con la API de AEMET y Entorno Seguro
* **`requests`**: Realizar peticiones HTTP robustas a los servidores y endpoints de la API de AEMET.
* **`urllib.parse`**: Formatear y codificar enlaces y parámetros de consulta de las URLs de forma correcta.
* **`dotenv.load_dotenv`**: Cargar de forma segura y privada tu `API Key` desde un archivo local `.env` sin exponer credenciales en el código.

### 🗄️ Base de Datos e Integración
* **`psycopg`**: Conectar y operar de forma nativa con bases de datos relacionales PostgreSQL para almacenar los históricos climatológicos de forma permanente.

### 🖥️ Visualización de Datos y Reportes
* **`IPython.display` (`Markdown`, `display`)**: Mostrar reportes enriquecidos, tablas con estilos HTML y formatos avanzados directamente dentro del entorno de Jupyter.


In [43]:
# Imports necesarios para el Proyecto AEMET 

#================================================================================#
#               SISTEMA OPERATIVO, RUTAS  Y ARCHIVOS                             # 
#================================================================================#
import os
from pathlib import Path
import zipfile
import sys
import subprocess
import gc
import pickle

#================================================================================#
#              CONTROL DE TIEMPO, FECHAS Y CALENDARIOS                           # 
#================================================================================#
import time
from time import sleep
import random
from datetime import datetime, timedelta
import calendar

#================================================================================#
#                 PROCESAMIENTO DE DATOS Y CÁLCULOS                              # 
#================================================================================#
import pandas as pd
import numpy as np
import json

#================================================================================#
#                CONEXIÓN WEB, APIS Y ENTORNO SEGURO                             # 
#================================================================================#
import requests
import urllib.parse
from dotenv import load_dotenv                  

#================================================================================#
#                      BASE DE DATOS E INTEGRACION                               # 
#================================================================================#
import psycopg    

#================================================================================#
#               INTERFAZ Y VISUALIZACIÓN (JUPYTER)                               # 
#================================================================================#
from IPython.display import Markdown, display   


## Paso 1: Configuración del Entorno y Carga de Variables Seguras

Para evitar exponer credenciales en el código fuente, la API Key de AEMET, de la conexions Postgree se almacena en el archivo `.env` de esta carpeta. Usamos la biblioteca `python-dotenv` para cargar el entorno de forma segura, apuntando específicamente al directorio de este notebook.

In [44]:
# Buscamos el archivo .env para la API KEY de AEMET y la contraseña de conexion de Postrgee
BASE_DIR = Path().resolve()
env_path = BASE_DIR / '.env'

# Carga del fichero .env para ver que tenemos conexion.
print("Buscando el archivo .env en:", env_path)
if env_path.exists():
    load_dotenv(dotenv_path=env_path)
    print("El archivo .env se ha cargado.")
else:
    print("Ojo: No se encontró el .env en esta carpeta. Probamos con el entorno general.")
    load_dotenv()

#================================================#
#          API KEY de AEMET Opendata             # 
#================================================#

# Sacamos la API Key
API_KEY = os.getenv('AEMET_API_KEY')
if not API_KEY:
    raise RuntimeError('No se encontró AEMET_API_KEY en las variables de entorno.')
else:
    # Mostramos un trozo por seguridad
    print(f"API Key encontrada: {API_KEY[:6]}...{API_KEY[-6:]}")

#================================================#
#          CONEXION A BD de POSTGREE             # 
#================================================#

# Sacamos la contraseña de Postgree 
PASS_OK = os.getenv('PASS_PG')
if not PASS_OK:
    raise RuntimeError('No se encontró PASS_original en las variables de entorno.')
else:
    # Mostramos un trozo por seguridad
    print(f"Key Postgree encontrada: {PASS_OK[:3]}...{PASS_OK[-6:]}")

# Sacamos puerto de Postgree
PORT_OK = os.getenv('PORT_PG')
if not PORT_OK:
    raise RuntimeError('No se encontró Puerto_original en las variables de entorno.')
else:
    # Mostramos un trozo por seguridad
    print(f"Port Postgree encontrada: {PORT_OK[:1]}...")

# Sacamos base de datos de Postgree
BD_OK = os.getenv('BD_PG')
if not BD_OK:
    raise RuntimeError('No se encontró base de datos en las variables de entorno.')
else:
    # Mostramos un trozo por seguridad
    print(f"Base de datos Postgree encontrada: {BD_OK[:1]}...{BD_OK[-1:]}")


Buscando el archivo .env en: C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\.env
El archivo .env se ha cargado.
API Key encontrada: eyJhbG...sjBWr4
Key Postgree encontrada: %40...gre%40
Port Postgree encontrada: 5...
Base de datos Postgree encontrada: a...t


---
### 🛠️ Funciones del Proyecto AEMET 🛠️ ###
---
### 📡 Conexión y Descarga de Datos (API OpenData)
* **`obtener_inventario_estaciones`**: Descargar el catálogo oficial de todas las estaciones meteorológicas disponibles en la API de AEMET, gestionando reintentos con esperas exponenciales (*backoff*) ante bloqueos por exceso de peticiones.
* **`obtener_valores_climaticos_todas`**: Solicitar y descargar los datos climáticos históricos diarios para todas las estaciones entre un rango de fechas específico en formato UTC, incorporando control robusto de fallos de red.
### 🧹 Limpieza, Conversión y Preparación
* **`aemet_coordenada_a_float`**: Traducir las coordenadas en formato sexagesimal de la API (ej. `402500N`, `034102W`) a floats decimales, aplicando el signo negativo si corresponde a las direcciones Sur u Oeste.
* **`limpiar_y_cargar_datos_fechas_todas_estaciones`**: Procesar el JSON de datos meteorológicos diarios para construir un DataFrame limpio, homogeneizar tipos de datos, ajustar decimales, corregir precipitaciones inapreciables (`"Ip"` a `0.0`) y parsear variables horarias.
* **`limpiar_y_cargar_datos_estaciones`**: Limpiar y estructurar el inventario maestro de estaciones en un DataFrame, asegurando la conversión de las coordenadas geográficas a float y la corrección de altitudes para la carga final en la base de datos.

In [66]:
#==================================================================================#
#  => Función para traer el listado de todas las estacionesde API AEMET OpenDAta   # 
#==================================================================================#
#    Por defecto max_retries = 10 y base_delay = 5.0 para intentos de conexion     #
#==================================================================================#

def obtener_inventario_estaciones( max_retries = 10, base_delay = 5.0):
    url_base = "https://opendata.aemet.es/opendata/api/valores/climatologicos/inventarioestaciones/todasestaciones"
    delay = base_delay
    print("Pidiendo el inventario a la API:", url_base)
    
    # Bucle de conexiones por si hay algun problema intenta 
    for intento in range(max_retries):
        
        try:

            response = requests.get(url_base, headers=headers, timeout=30)
            
            # Si no tenemos permiso o no existe el endpoint, paramos
            # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)
            if response.status_code in [401, 403, 404]:
                print(f"Error grave {response.status_code}, no se puede reintentar.")
                return None

            # Como ha dado 200 es ok la conexion
            if response.status_code == 200:
                meta_datos = response.json()
                estado = meta_datos.get('estado')
                print(f"Respuesta API: {meta_datos.get('descripcion')} (Estado: {meta_datos.get('estado')})")
                
                if meta_datos.get('estado') == 200:
                    url_datos = meta_datos.get('datos')
                    print("Tenemos enlace temporal, descargando los datos reales...")

                if estado == 200:
                    # Con la URL url_final de datos nos conectamos otra vez para que nos de los datos 
                    url_final = meta_datos.get('datos')
                    datos_response = requests.get(url_final, timeout=30)
                    # Si conexion OK nos da el JSON con los datos con un timeout 30 segundos por sino responde     
                    if datos_response.status_code == 200:
                        return datos_response.json()
                    # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)
                    elif datos_response.status_code in [401, 403, 404]:
                        print(f"Error grave en datos {datos_response.status_code}, no reintentamos.")
                        return None
                    else:
                        print(f"Intento {intento+1} fallido al descargar datos: {datos_response.status_code}")
                # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)        
                elif estado in [401, 403, 404]:
                    print(f"Error API: Estado {estado}: {meta_datos.get('descripcion')}. No se reintentará.")
                    return None
                # si la conexion da Codigos de error : 429 Too Many Requests (superado el limite de peticiones)    
                elif estado == 429:
                    print(f"Intento {intento+1}: La API dice Estado 429 (Límite superado).")
                else:
                    print(f"Intento {intento+1}: API devolvió Estado {estado}: {meta_datos.get('descripcion')}")
            # si la conexion da Codigos de error : 429 Too Many Requests (superado el limite de peticiones)
            elif response.status_code == 429:
                print(f"Intento {intento+1}: HTTP 429 (Muchas peticiones).")
            else:
                print(f"Intento {intento+1}: Código de red {response.status_code}")

        #salida por maximo intentos de conexion        
        except Exception as e:
            print(f"Intento {intento+1}: Error de red/conexión: {e}")

        # sigue intentandolo hastq eue llegue al maximo intentos             
        if intento < max_retries - 1:
            print(f"Esperando {delay} segundos antes de reintentar...")
            time.sleep(delay)
            delay *= 2.0

    print(f"Funcion [obtener_inventario_estaciones] No se pudo descargar el bloque tras {max_retries} intentos.")        
    return None

#==================================================================================================#
#  => Función para traer el JSON de todas las estaciones de API AEMET OpenDAta  entre dos fechas   # 
#==================================================================================================#
#   .-  Por defecto max_retries = 15 y base_delay = 5.0 para intentos de conexion                  #
#   .-  fechaini = fecha_ini_utc (formato especial de la API 2026-05-01T00:00:00UTC                #
#   .-  fechafin = fecha_ini_utc (formato especial de la API 2026-05-15T00:00:00UTC                #
#==================================================================================================#

# Petición para todas las estaciones con reintentos y esperas (backoff) si falla la API
def obtener_valores_climaticos_todas(fecha_ini_utc, fecha_fin_utc, max_retries=15, base_delay=5.0):
    url_base = f"https://opendata.aemet.es/opendata/api/valores/climatologicos/diarios/datos/fechaini/{fecha_ini_utc}/fechafin/{fecha_fin_utc}/todasestaciones"
    delay = base_delay

    # Bucle de conexiones por si hay algun problema intenta 
    for intento in range(max_retries):

        try:

            response = requests.get(url_base, headers=headers, timeout=30)
            
            # Si no tenemos permiso o no existe el endpoint, paramos
            # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)
            if response.status_code in [401, 403, 404]:
                print(f"Error grave {response.status_code}, no se puede reintentar.")
                return None

            # Como ha dado 200 es ok la conexion                
            if response.status_code == 200:
                meta_datos = response.json()
                estado = meta_datos.get('estado')
                
                if estado == 200:
                    # Con la URL url_final de datos nos conectamos otra vez para que nos de los datos                     
                    url_final = meta_datos.get('datos')
                    datos_response = requests.get(url_final, timeout=30)
                    # Si conexion OK nos da el JSON con los datos con un timeout 30 segundos por sino responde                   
                    if datos_response.status_code == 200:
                        return datos_response.json()
                    # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)
                    elif datos_response.status_code in [401, 403, 404]:
                        print(f"Error grave en datos {datos_response.status_code}, no reintentamos.")
                        return None
                    else:
                        print(f"Intento {intento+1} fallido al descargar datos: {datos_response.status_code}")
                # Codigos de error : 401 Unauthorized (No autorizado), 403 Forbidden (Prohibido), 404 Not Found (No encontrado)        
                elif estado in [401, 403, 404]:
                    print(f"Error API: Estado {estado}: {meta_datos.get('descripcion')}. No se reintentará.")
                    return None
                # si la conexion da Codigos de error : 429 Too Many Requests (superado el limite de peticiones)    
                elif estado == 429:
                    print(f"Intento {intento+1}: La API dice Estado 429 (Límite superado).")
                else:
                    print(f"Intento {intento+1}: API devolvió Estado {estado}: {meta_datos.get('descripcion')}")
            # si la conexion da Codigos de error : 429 Too Many Requests (superado el limite de peticiones)    
            elif response.status_code == 429:
                print(f"Intento {intento+1}: HTTP 429 (Muchas peticiones).")
            else:
                print(f"Intento {intento+1}: Código de red {response.status_code}")
        # salida por maximo intentos de conexion                 
        except Exception as e:
            print(f"Intento {intento+1}: Error de red/conexión: {e}")

        # sigue intentandolo hastq eue llegue al maximo intentos 
        if intento < max_retries - 1:
            print(f"Esperando {delay} segundos antes de reintentar...")
            time.sleep(delay)
            delay *= 2.0
            
    print(f"No se pudo descargar el bloque tras {max_retries} intentos.")
    return None


#================================================================================#
#  => Funcion para traducir latitud/longitud ("402500N","034102W") en un Float   # 
#================================================================================#
def aemet_coordenada_a_float(texto_coord):
    
    # Si el dato viene vacio o no es texto, devolvemos NaN (not a number)
    if pd.isna(texto_coord) or not isinstance(texto_coord, str):
        return None
    
    # 1. Quitamos espacios por si acaso
    texto_coord = texto_coord.strip()
    
    # 2. Separamos la letra del final (N, S, E, W) y los numeros
    letra = texto_coord[-1].upper()   # Coge el ultimo carácter
    numeros = texto_coord[:-1]        # Coge todo menos el ultimo carácter
    
    # 3. Extraemos Grados, Minutos y Segundos troceando el texto
    # Rellenamos con ceros a la izquierda si falta algun digito (debe tener 6 digitos)
    numeros = numeros.zfill(6) 
    
    # => Grados, los dos primeros digitos
    grados = float(numeros[0:2])
    # => Minutos, los dos del medio    
    minutos = float(numeros[2:4]) 
    # => Segundos,los dos ultimos  
    segundos = float(numeros[4:6])
    
    # 4. Aplicamos la formula matematica
    decimal = grados + (minutos / 60.0) + (segundos / 3600.0)
    
    # 5. Si es Sur o Oeste, el numero es negativo
    if letra in ['S', 'W']:
        decimal = -decimal
        
    return decimal


#==================================================================================================#
#  => Función para limpiar el JSON de todas las estaciones de API AEMET OpenDAta entre dos fechas  # 
#==================================================================================================#
#   .- Asjustamos los datos para la carga 
#==================================================================================================#

# Limpieza y preparación de los datos obtenidos
def limpiar_y_cargar_datos_fechas_todas_estaciones(datos_json):
    if not datos_json:
        return pd.DataFrame()
        
    datos_lista = []
    for dia in datos_json:
        datos_lista.append({ "fecha"      : dia.get("fecha"),
                             "indicativo" : dia.get("indicativo"),
                             "nombre"     : dia.get("nombre"),
                             "provincia"  : dia.get("provincia"),
                             "altitud"    : dia.get("altitud"),
                             "tmed"       : dia.get("tmed"),
                             "prec"       : dia.get("prec"),
                             "tmin"       : dia.get("tmin"),
                             "horatmin"   : dia.get("horatmin"),
                             "tmax"       : dia.get("tmax"),
                             "horatmax"   : dia.get("horatmax"),
                             "dir"        : dia.get("dir"),
                             "velmedia"   : dia.get("velmedia"),
                             "racha"      : dia.get("racha"),
                             "horaracha"  : dia.get("horaracha"),
                             "sol"        : dia.get("sol"),
                             "presMax"    : dia.get("presMax"),
                             "horaPresMax": dia.get("horaPresMax"),
                             "presMin"    : dia.get("presMin"),
                             "horaPresMin": dia.get("horaPresMin"),
                             "hrMedia"    : dia.get("hrMedia"),
                             "hrMax"      : dia.get("hrMax"),
                             "horaHrMax"  : dia.get("horaHrMax"),
                             "hrMin"      : dia.get("hrMin"),
                             "horaHrMin"  : dia.get("horaHrMin")
                           })

    df = pd.DataFrame(datos_lista)

    # 1. Ajustes basicos de tipos de datos
    df["fecha"]      = pd.to_datetime(df["fecha"])
    df["altitud"]    = pd.to_numeric(df["altitud"], errors="coerce")
    df["indicativo"] = df["indicativo"].astype(str)
    df["nombre"]     = df["nombre"].astype(str)
    df["provincia"]  = df["provincia"].astype(str)
    df["dir"]        = df["dir"].astype(str)
    df["horaracha"]  = df["horaracha"].astype(str)

    # Funcion para pasar numeros con coma a float con punto
    def a_float(columna):
        if columna is None:
            return np.nan
        return columna.astype(str).str.replace(",", ".").replace("", "nan").astype(np.float32)

    # 2. Convertimos las columnas numericas
    columnas_numericas = ["tmed", "tmin", "tmax", "velmedia", "racha", "sol", "presMax", "presMin", "hrMedia", "hrMax", "hrMin"]
    for col in columnas_numericas:
        if col in df.columns:
            df[col] = a_float(df[col])

    # 3. Arreglamos la precipitacion (la API a veces devuelve "Ip" que significa inapreciable, la ponemos en 0.0)
    if "prec" in df.columns:
        df["prec"] = df["prec"].astype(str).str.replace("Ip", "0.0").str.replace(",", ".").replace("", "0.0")
        df["prec"] = pd.to_numeric(df["prec"], errors="coerce").astype(np.float32)

    # 4. Limpiamos horas y las pasamos a tipo time de python
    columnas_tiempo = ["horatmin", "horatmax", "horaHrMax", "horaHrMin"]
    for col in columnas_tiempo:
        if col in df.columns:
            df[col] = df[col].astype(str).replace(["Varias", "nan", "None", ""], np.nan)
            df[col] = pd.to_datetime(df[col], format="%H:%M", errors="coerce").dt.time

    # 5. Otras horas que vienen solo con la hora
    columnas_hora_sola = ["horaPresMax", "horaPresMin"]
    for col in columnas_hora_sola:
        if col in df.columns:
            df[col] = df[col].astype(str).replace(["Varias", "nan", "None", ""], np.nan)
            df[col] = pd.to_datetime(df[col], format="%H", errors="coerce").dt.time

    return df

#==================================================================================================#
#  => Función para limpiar el JSON del maestro  todas las estaciones de API AEMET OpenDAta         # 
#==================================================================================================#
#   .- Ajustamos los datos para la carga;                                                          #
#      . Caso especifico de limpieza es la conversion de la latitud y longitud con la funcion      #
#        "def aemet_coordenada_a_float()"  Ejemplo de como viene el dato [394924N , 025309E]       #
#==================================================================================================#

def limpiar_y_cargar_datos_estaciones(datos_json):
    
     data = []
     for estacion in estaciones:

        # Mapeamos los datos de JSON para hacer un diccionario de carga
        data.append({ 'indicativo': estacion["indicativo"],
                      'nombre'    : estacion["nombre"],
                      'provincia' : estacion["provincia"],
                      'latitud'   : estacion["latitud"],
                      'longitud'  : estacion["longitud"],
                      'altitud'   : estacion["altitud"],
                      'indsinop'  : estacion["indsinop"]
                     })

     df = pd.DataFrame(data)

     # Aplicamos la conversión a las columnas del DataFrame para el caso de la latitud y longitud
     df['latitud']  = df['latitud'].apply(aemet_coordenada_a_float).astype(np.float32)
     df['longitud'] = df['longitud'].apply(aemet_coordenada_a_float).astype(np.float32)

     # Aseguramos que la altitud también sea numérica (por si viene como string) si viene vacio nan , errors='coerce'
     df['altitud'] = pd.to_numeric(df['altitud'], errors='coerce')
    
     return df


## Paso 2: El Protocolo de Conexión de AEMET OpenData

La API de AEMET funciona bajo un modelo de **"doble llamada"**:
1. Enviamos nuestra consulta con la clave `api_key` en las cabeceras (headers) de la solicitud HTTP GET.
2. AEMET valida la solicitud y nos responde con un JSON de metadatos que contiene una URL temporal en el campo `datos`.
3. Hacemos una segunda solicitud `GET` a esa URL temporal (esta vez sin cabeceras de autorización) para descargar los datos reales del sensor en formato JSON.

Definimos el diccionario de cabeceras que utilizaremos en la primera llamada de autenticación.

In [59]:
# Preparamos las cabeceras para la petición
headers = {
    'cache-control': "no-cache",
    'api_key': API_KEY,
    'accept': "application/json"
}

print("Cabeceras listas para usar.")

Cabeceras listas para usar.


## Paso 3: Obtener el Inventario Completo de Estaciones Climatológicas

Para realizar cualquier consulta histórica, necesitamos conocer el identificador único (`indicativo` o `idema`) de la estación meteorológica. El endpoint `/api/valores/climatologicos/inventarioestaciones/todasestaciones` nos devuelve el catálogo completo de las estaciones activas en España.

In [60]:
# Probamos a traer las estaciones
estaciones = obtener_inventario_estaciones()

if estaciones:
    print(f"Total estaciones cargadas: {len(estaciones)}")
    df_estaciones = pd.DataFrame(estaciones)
    print("Primeras 5 estaciones:")
    
else:
    print("No se pudo obtener el inventario. Revisa la clave API.")
df_estaciones.head(5)    

Pidiendo el inventario a la API: https://opendata.aemet.es/opendata/api/valores/climatologicos/inventarioestaciones/todasestaciones
Intento 1: HTTP 429 (Muchas peticiones).
Esperando 5.0 segundos antes de reintentar...
Intento 2: HTTP 429 (Muchas peticiones).
Esperando 10.0 segundos antes de reintentar...
Respuesta API: exito (Estado: 200)
Tenemos enlace temporal, descargando los datos reales...
Total estaciones cargadas: 920
Primeras 5 estaciones:


,latitud,provincia,altitud,indicativo,nombre,indsinop,longitud
0,394924N,ILLES BALEARS,490,B013X,"ESCORCA, LLUC",08304,025309E
1,394744N,BALEARES,5,B051A,"SÓLLER, PUERTO",08316,024129E
2,394121N,ILLES BALEARS,60,B087X,BANYALBUFAR,,023046E
3,393446N,BALEARES,52,B103B,ANDRATX - SANT ELM,,022208E
4,393305N,BALEARES,50,B158X,"CALVIÀ, ES CAPDELLÀ",,022759E


In [61]:
# Aplicamos la limpieza de datos y la transformacion.(latitud y longitud)
df_estaciones = limpiar_y_cargar_datos_estaciones(estaciones)

# Mostramos las primeras 5 filas
df_estaciones.head(5) 

,indicativo,nombre,provincia,latitud,longitud,altitud,indsinop
0,B013X,"ESCORCA, LLUC",ILLES BALEARS,39.823334,2.885833,490,08304
1,B051A,"SÓLLER, PUERTO",BALEARES,39.795555,2.691389,5,08316
2,B087X,BANYALBUFAR,ILLES BALEARS,39.689167,2.512778,60,
3,B103B,ANDRATX - SANT ELM,BALEARES,39.579445,2.368889,52,
4,B158X,"CALVIÀ, ES CAPDELLÀ",BALEARES,39.551388,2.466389,50,


## Paso 4: Consulta de Prueba de Valores Climatológicos Diarios

Para verificar el flujo de descarga de datos de sensores diarios, consultaremos los datos del 1 al 15 de mayo de 2026 para todas las estaciones (`idemas`).

El formato del endpoint requerido es:
`/api/valores/climatologicos/diarios/datos/fechaini/{fecha_ini_utc}/fechafin/{fecha_fin_utc}/todasestaciones`

In [64]:
# Hacemos una prueba en la primera quincena de Mayo 2026
ini_prueba = "2026-05-01T00:00:00UTC"
fin_prueba = "2026-05-15T00:00:00UTC"

print("Probando descarga para todas las estaciones primera quincena de mayo 2026...")
climaticos_prueba = obtener_valores_climaticos_todas(ini_prueba, fin_prueba)

if climaticos_prueba:
    print(f"¡Funciona! Hemos descargado {len(climaticos_prueba)} filas diarias.")
    print("Ejemplo del primer registro en crudo:")
    
    print(json.dumps(climaticos_prueba[0], indent=2, ensure_ascii=False))
else:
    print("No se han recibido datos. Revisa fechas y tu API Key.")

Probando descarga para todas las estaciones primera quincena de mayo 2026...
¡Funciona! Hemos descargado 12239 filas diarias.
Ejemplo del primer registro en crudo:
{
  "fecha": "2026-05-01",
  "indicativo": "8293X",
  "nombre": "XÀTIVA",
  "provincia": "VALENCIA",
  "altitud": "88",
  "tmed": "18,2",
  "prec": "2,0",
  "tmin": "16,1",
  "horatmin": "23:20",
  "tmax": "20,4",
  "horatmax": "13:10",
  "dir": "36",
  "velmedia": "1,7",
  "racha": "6,4",
  "horaracha": "17:00",
  "presMax": "1012,9",
  "horaPresMax": "21",
  "presMin": "1008,2",
  "horaPresMin": "04",
  "hrMedia": "82",
  "hrMax": "96",
  "horaHrMax": "Varias",
  "hrMin": "74",
  "horaHrMin": "17:10"
}


## Paso 5: Procesamiento y Limpieza de Datos (Pipeline ETL)

Los datos crudos de AEMET vienen completamente en formato de texto (strings). Además, la representación decimal utiliza la coma española (`,`) en lugar del punto (`.`), lo cual impide operaciones aritméticas en Pandas.

Implementamos la función `limpiar_y_cargar_datos` que realiza el siguiente flujo de transformación:
1. Extrae los campos de interés de cada registro.
2. Convierte el campo `fecha` a tipo `datetime` de Pandas.
3. Normaliza las representaciones numéricas: reemplaza las comas por puntos en variables continuas y las convierte a tipo `float32` o `int` de forma segura.
4. Convierte las horas de temperaturas/presiones máximas y mínimas a objetos `datetime.time`, gestionando de forma robusta valores de texto anómalos como 'Varias' o nulos.
5. Limpia el campo de precipitación (`prec`), convirtiendo el valor especial "Ip" (inapreciable) a `0.0` para poder realizar cálculos y sumas acumuladas consistentes.

In [69]:
# Aplicamos la limpieza de datos y la transformacion.(latitud y longitud)
df_fechas_estaciones_final = limpiar_y_cargar_datos_fechas_todas_estaciones(climaticos_prueba)

# Mostramos las primeras 5 filas
df_fechas_estaciones_final.head(15) 


,fecha,indicativo,nombre,provincia,altitud,tmed,prec,tmin,horatmin,tmax,...,sol,presMax,horaPresMax,presMin,horaPresMin,hrMedia,hrMax,horaHrMax,hrMin,horaHrMin
0,2026-05-01,8293X,XÀTIVA,VALENCIA,88,18.200001,2.0,16.100000,23:20:00,20.400000,...,NaN,1012.900024,21:00:00,1008.200012,04:00:00,82.0,96.0,NaT,74.0,17:10:00
1,2026-05-01,8270X,BICORP,VALENCIA,305,16.600000,0.6,15.100000,04:15:00,18.000000,...,NaN,NaN,NaT,NaN,NaT,92.0,96.0,NaT,86.0,12:20:00
2,2026-05-01,8492X,ATZENETA DEL MAESTRAT,CASTELLON,420,16.000000,0.4,14.500000,05:20:00,17.500000,...,NaN,NaN,NaT,NaN,NaT,83.0,92.0,NaT,76.0,09:10:00
3,2026-05-01,5612B,LA RODA DE ANDALUCÍA,SEVILLA,410,18.000000,0.0,9.200000,04:00:00,26.799999,...,NaN,971.400024,21:00:00,969.400024,02:00:00,62.0,95.0,06:30:00,37.0,11:50:00
4,2026-05-01,7250C,ABANILLA,MURCIA,174,21.100000,0.4,16.700001,NaT,25.500000,...,NaN,NaN,NaT,NaN,NaT,75.0,92.0,07:00:00,55.0,13:50:00
5,2026-05-01,3094B,TARANCÓN,CUENCA,819,19.299999,0.0,12.400000,06:00:00,26.200001,...,NaN,927.799988,NaT,925.099976,02:00:00,51.0,83.0,06:10:00,13.0,16:10:00
6,2026-05-01,2462,PUERTO DE NAVACERRADA,MADRID,1892,10.300000,2.2,5.700000,05:20:00,14.900000,...,12.3,815.400024,21:00:00,813.099976,02:00:00,62.0,94.0,23:50:00,47.0,15:40:00
7,2026-05-01,9302Y,TUDELA,NAVARRA,297,18.900000,0.0,14.700000,03:26:00,23.100000,...,NaN,NaN,NaT,NaN,NaT,73.0,88.0,NaT,59.0,NaT
8,2026-05-01,2755X,BENAVENTE,ZAMORA,715,15.600000,1.6,9.700000,05:50:00,21.500000,...,10.5,936.700012,09:00:00,934.799988,17:00:00,58.0,95.0,05:20:00,34.0,14:30:00
9,2026-05-01,1679A,MONFORTE DE LEMOS,LUGO,291,17.600000,0.0,11.700000,05:28:00,23.600000,...,NaN,NaN,NaT,NaN,NaT,69.0,94.0,06:40:00,46.0,16:20:00


## Paso 6: Extracción Histórica Masiva con Control de Rate Limiting

Para obtener un dataset consistente e histórico, se implementa un bucle recursivo que descarga datos año por año y mes por mes.

**⚠️ IMPORTANTE - Límites de la API de AEMET:**
AEMET limita estrictamente la cantidad de llamadas simultáneas. Si no se manejan pausas, la API responderá con errores `HTTP 429` (Demasiadas peticiones) o bloqueará el token de usuario.
Para evitar esto, implementamos:
1. **Pausa de 60 segundos** entre cada año consultado.
2. **Pausa de 1.5 segundos** entre cada mes consultado.
3. **Pausa de 40 minutos** de seguridad si el tiempo total de ejecución supera los 5 minutos de forma continua.
4. **Saltado dinámico de meses futuros** para evitar llamadas inútiles sobre periodos que aún no han ocurrido.

*Nota de control:* Por defecto, esta descarga está pre-configurada para extraer los **últimos 3 años** (2024-2026) para optimizar el tiempo de prueba de este notebook. Si deseas obtener la serie histórica completa de los últimos 10 años, simplemente modifica la variable `AÑOS_A_CONSULTAR = 10`.

In [75]:
# Ajustamos las fechas para la descarga histórica (10 años en bloques)
fecha_fin_global = datetime(2026, 6, 13)
DIAS_A_CONSULTAR = 3650  # 10 años
fecha_inicio_global = fecha_fin_global - timedelta(days=DIAS_A_CONSULTAR - 1)

# Carpetas de salida
OUTPUT_CSV_DIR = BASE_DIR / 'sheets' / 'csv'
OUTPUT_XLSX_DIR = BASE_DIR / 'sheets' / 'xlsx'
OUTPUT_CSV_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_XLSX_DIR.mkdir(parents=True, exist_ok=True)

año_inicio = fecha_inicio_global.year
año_fin = fecha_fin_global.year

print(f"Empezando la descarga masiva por años desde {año_inicio} hasta {año_fin}...")

for año in range(año_fin, año_inicio - 1, -1):
    print("\n-----------------------------------------")
    print(f"Procesando el año: {año}")
    print("-----------------------------------------")
    
    ini_año = max(datetime(año, 1, 1), fecha_inicio_global)
    fin_año = min(datetime(año, 12, 31), fecha_fin_global)
    
    df_año = pd.DataFrame()
    fecha_actual = fin_año
    
    lote_size = 15  # Descargamos en bloques de 15 días
    dias_año_procesados = 0
    total_dias_año = (fin_año - ini_año).days + 1
    
    print(f"Fechas para {año}: de {ini_año.strftime('%Y-%m-%d')} a {fin_año.strftime('%Y-%m-%d')} ({total_dias_año} días)")
    
    while dias_año_procesados < total_dias_año:
        dias_restantes = total_dias_año - dias_año_procesados
        dias_lote = min(lote_size, dias_restantes)
        
        fecha_ini_lote = fecha_actual - timedelta(days=dias_lote - 1)
        
        fecha_ini_str = fecha_ini_lote.strftime("%Y-%m-%dT00:00:00UTC")
        fecha_fin_str = fecha_actual.strftime("%Y-%m-%dT23:59:59UTC")
        
        print(f"Descargando bloque: {fecha_ini_str} a {fecha_fin_str}...")
        
        # Esperamos 2 segundos entre descargas para no pasarnos del límite de la API
        print("Pausa de 2 segundos...")
        time.sleep(2.0)
        
        datos_json = obtener_valores_climaticos_todas(fecha_ini_str, fecha_fin_str)
        
        if datos_json:
            df_lote = limpiar_y_cargar_datos_fechas_todas_estaciones(datos_json)
            df_año = pd.concat([df_año, df_lote], ignore_index=True)
            print(f"Ok, bloque descargado. Registros nuevos: {len(df_lote)}")
        else:
            print(f"Error en bloque de {fecha_ini_str} a {fecha_fin_str}")
            
        dias_año_procesados += dias_lote
        fecha_actual = fecha_ini_lote - timedelta(days=1)
        
    # Guardamos el año completo y limpiamos memoria RAM
    if not df_año.empty:
        # Ordenar por fecha cronológica más antigua a más reciente
        if 'fecha' in df_año.columns:
            df_año = df_año.sort_values(by='fecha', ascending=True)
            
        csv_path = OUTPUT_CSV_DIR / f'climaticos_{año}.csv'
        excel_path = OUTPUT_XLSX_DIR / f'climaticos_{año}.xlsx'
        
        print(f"Guardando ficheros para el año {año}...")
        # Exportar el dataset a un CSV para ver el contenido .
        try:
            df_año.to_csv(csv_path, index=False) # index=False para no incluir la columna de indices en el archivo exportado.   
            print(f"Guardado CSV en: {csv_path}")
        except Exception as e:
            print(f"Error al guardar los archivos de {año}: {e}")

        # Exportar el dataset a un xlsx para ver el contenido .
        try:
            df_año.to_excel(excel_path, index=False) # index=False para no incluir la columna de indices en el archivo exportado.   
            print(f"Guardado Excel en: {excel_path}")        
        except Exception as e:
            print(f"Error al guardar los archivos de {año}: {e}")           
                      
        # Forzamos la limpieza de RAM
        del df_año
        gc.collect()
    else:
        print(f"No hay datos para el año {año}, no guardamos nada.")

print("\n¡Descarga y exportación completadas!")

Empezando la descarga masiva por años desde 2016 hasta 2026...

-----------------------------------------
Procesando el año: 2026
-----------------------------------------
Fechas para 2026: de 2026-01-01 a 2026-06-13 (164 días)
Descargando bloque: 2026-05-30T00:00:00UTC a 2026-06-13T23:59:59UTC...
Pausa de 2 segundos...
Intento 1: HTTP 429 (Muchas peticiones).
Esperando 5.0 segundos antes de reintentar...
Intento 2: HTTP 429 (Muchas peticiones).
Esperando 10.0 segundos antes de reintentar...
Ok, bloque descargado. Registros nuevos: 10556
Descargando bloque: 2026-05-15T00:00:00UTC a 2026-05-29T23:59:59UTC...
Pausa de 2 segundos...
Ok, bloque descargado. Registros nuevos: 12273
Descargando bloque: 2026-04-30T00:00:00UTC a 2026-05-14T23:59:59UTC...
Pausa de 2 segundos...
Ok, bloque descargado. Registros nuevos: 12241
Descargando bloque: 2026-04-15T00:00:00UTC a 2026-04-29T23:59:59UTC...
Pausa de 2 segundos...
Ok, bloque descargado. Registros nuevos: 12380
Descargando bloque: 2026-03-31T00

## Paso 7: Verificación de los Archivos Físicos Generados

Para validar que la exportación anual y segmentada se ha completado correctamente en disco, en esta sección listamos el contenido de los directorios de salida  y , mostrando el tamaño ocupado por cada archivo generado.

In [77]:
# Comprobamos los archivos creados en el disco y su tamaño
OUTPUT_CSV_DIR = BASE_DIR / 'sheets' / 'csv'
OUTPUT_XLSX_DIR = BASE_DIR / 'sheets' / 'xlsx'

print("Archivos CSV creados:")
if OUTPUT_CSV_DIR.exists():
    for f in sorted(OUTPUT_CSV_DIR.glob('climaticos_*')):
        size_mb = f.stat().st_size / (1024 * 1024)
        print(f" - csv/{f.name} ({size_mb:.2f} MB)")
else:
    print("El directorio de CSVs no existe.")

print("\nArchivos Excel creados:")
if OUTPUT_XLSX_DIR.exists():
    for f in sorted(OUTPUT_XLSX_DIR.glob('climaticos_*')):
        size_mb = f.stat().st_size / (1024 * 1024)
        print(f" - xlsx/{f.name} ({size_mb:.2f} MB)")
else:
    print("El directorio de Excels no existe.")

Archivos CSV creados:
 - csv/climaticos_2016.csv (20.90 MB)
 - csv/climaticos_2017.csv (38.87 MB)
 - csv/climaticos_2018.csv (38.68 MB)
 - csv/climaticos_2019.csv (39.21 MB)
 - csv/climaticos_2020.csv (39.20 MB)
 - csv/climaticos_2021.csv (39.78 MB)
 - csv/climaticos_2022.csv (40.05 MB)
 - csv/climaticos_2023.csv (40.49 MB)
 - csv/climaticos_2024.csv (40.88 MB)
 - csv/climaticos_2025.csv (40.60 MB)
 - csv/climaticos_2026.csv (17.47 MB)

Archivos Excel creados:
 - xlsx/climaticos_2016.xlsx (20.91 MB)
 - xlsx/climaticos_2017.xlsx (38.89 MB)
 - xlsx/climaticos_2018.xlsx (38.86 MB)
 - xlsx/climaticos_2019.xlsx (39.22 MB)
 - xlsx/climaticos_2020.xlsx (39.24 MB)
 - xlsx/climaticos_2021.xlsx (39.78 MB)
 - xlsx/climaticos_2022.xlsx (39.98 MB)
 - xlsx/climaticos_2023.xlsx (40.36 MB)
 - xlsx/climaticos_2024.xlsx (40.79 MB)
 - xlsx/climaticos_2025.xlsx (40.52 MB)
 - xlsx/climaticos_2026.xlsx (17.37 MB)


## ℹ️ Nota sobre la Variabilidad en el Número de Estaciones (Idemas) por Lote

Es común observar que, aunque el inventario maestro de AEMET reporta aproximadamente **920 estaciones activas**, el número de registros diarios reales por lote no equivale exactamente a `920 estaciones × N días`.

Por ejemplo, en un lote de 15 días (como del 16 al 30 de mayo), se obtienen unos **12,176 registros**, lo que equivale a un promedio de **~812 estaciones diarias** con reporte de datos, en lugar de las 920 teóricas.

### ¿A qué se debe esta diferencia?
1. **Estaciones inactivas o en mantenimiento:** Algunas estaciones pueden sufrir fallos de comunicación, averías en sensores o estar en periodos de mantenimiento programado.
2. **Validación y desfase temporal de datos:** AEMET somete los datos climatológicos a controles de calidad antes de publicarlos en la API OpenData. Las estaciones de publicación reciente o secundarias pueden tardar más en volcar sus datos.
3. **Altas y bajas en la red:** El inventario incluye estaciones históricas o de alta reciente que no necesariamente reportan en la horquilla temporal consultada.
4. **Filtros de datos nulos:** Durante el proceso de ingesta, aquellos registros que no contienen lecturas válidas o están totalmente vacíos para los días del lote son descartados o no transmitidos por la propia API.

## Paso 8: **Creacion de tablas:** 

Crear tablas para Postgree.


In [ ]:
# CREATE DATABASE aemet;

# CREATE TABLE IF NOT EXISTS estaciones (
#                                        indicativo VARCHAR (10) PRIMARY KEY,
#                                        nombre     VARCHAR (150),
#                                        provincia  VARCHAR (100),
#                                        latitud    FLOAT,
#                                        longitud   FLOAT,
#                                        altitud    INT,
#                                        indsinop   VARCHAR (10)  
#                                        );

# CREATE TABLE IF NOT EXISTS datos_climaticos (
#                                              fechA       DATE,
#                                              indicativo  VARCHAR (10) ,
#                                              nombre      VARCHAR(150),
#                                              provincia   VARCHAR(255),
#                                              altitud     INT,
#                                              tmed        FLOAT,
#                                              prec        FLOAT,
#                                              tmin        FLOAT,
#                                              horatmin    TIME,
#                                              tmax        FLOAT,
#                                              horatmax    TIME,
#                                              dir         FLOAT,
#                                              velmedia    FLOAT,
#                                              racha       FLOAT,
#                                              horaracha   VARCHAR(15),
#                                              sol         FLOAT,
#                                              presMax     FLOAT,
#                                              horaPresMax TIME,
#                                              presMin     FLOAT,
#                                              horaPresMin TIME,
#                                              hrMedia     FLOAT,
#                                              hrMax       FLOAT,
#                                              horaHrMax   TIME,
#                                              hrMin       FLOAT,
#                                              horaHrMin   TIME,
#                                              PRIMARY KEY (indicativo, fecha)
#                                             );


## Paso 9: **Conexion y carga maestro a Postgree:** 

Conexion y carga de los datos de las estaciones como maestro.

In [ ]:
# Conexion al Postgree a la aemet
DATABASE_URI: str = f"postgresql://postgres:{PASS_OK}@localhost:{PORT_OK}/{BD_OK}"

In [80]:
# Probamos a traer las estaciones
estaciones = obtener_inventario_estaciones()

if estaciones:
    print(f"Total estaciones cargadas: {len(estaciones)}")
    df_estaciones = pd.DataFrame(estaciones)  
else:
    print("No se pudo obtener el inventario. Revisa la clave API.")

# Aplicamos la limpieza de datos y la transformacion.(latitud y longitud)
df_estaciones = limpiar_y_cargar_datos_estaciones(estaciones)


Pidiendo el inventario a la API: https://opendata.aemet.es/opendata/api/valores/climatologicos/inventarioestaciones/todasestaciones
Respuesta API: exito (Estado: 200)
Tenemos enlace temporal, descargando los datos reales...
Total estaciones cargadas: 920


In [81]:

# Conectamos a Postgree y caragamos el maestro de estaciones
with psycopg.connect(DATABASE_URI) as conn:
     with conn.cursor() as cursor:
        
        # Talbla estaciones de aemet
         data_estaciones = df_estaciones.replace({np.nan : None}).values.tolist() 
         cursor.executemany(query = """INSERT INTO estaciones ( indicativo , nombre , provincia , latitud , longitud , altitud , indsinop )
                                                       VALUES (%s, %s, %s, %s, %s, %s, %s)
                                                       ON CONFLICT (indicativo) DO NOTHING;;""",
                            params_seq =  data_estaciones)


## Paso 10: **Conexion y carga datos estaciones entre fechas a Postgree:** 
Conexion y carga de los datos de las estaciones entre fechas.

In [82]:
# Conexion al Postgree a la aemet
DATABASE_URI: str = f"postgresql://postgres:{PASS_OK}@localhost:{PORT_OK}/{BD_OK}"

In [98]:
# Ajustamos las fechas para la descarga histórica (10 años en bloques)
fecha_fin_global = datetime(2026, 6, 13)
DIAS_A_CONSULTAR = 3650  # 10 años
fecha_inicio_global = fecha_fin_global - timedelta(days=DIAS_A_CONSULTAR - 1)

# Carpetas de salida
OUTPUT_JSON_DIR = BASE_DIR / 'sheets' / 'json'
OUTPUT_JSON_DIR.mkdir(parents=True, exist_ok=True)

año_inicio = fecha_inicio_global.year
año_fin = fecha_fin_global.year

print(f"Empezando la descarga masiva por años desde {año_inicio} hasta {año_fin}...")

for año in range(año_inicio, año_fin + 1):

    csv_path = OUTPUT_CSV_DIR / f"climaticos_{año}.csv"
    print (csv_path) 
    
    # cargamos el csv de cada año directamente
    df_año = pd.read_csv(csv_path)    

    # lo convertimos a json en formato de registros
    json_data = df_año.to_json(orient='records', force_ascii=False)
    
    # lo formateamos para que sea legible por humanos (comentar estas dos lineas si prefieres ahorrar espacio)
    json_data = json.dumps(json.loads(json_data), indent=4, ensure_ascii=False)

    # guardamos cada año en un archivo independiente
    json_path = OUTPUT_JSON_DIR / f"climaticos_{año}.json"
    print (json_path)
    with open(f"{json_path}", "w", encoding="utf-8") as file:
         file.write(json_data)     
        


Empezando la descarga masiva por años desde 2016 hasta 2026...
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2016.csv
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2016.json
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2017.csv
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2017.json
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2018.csv
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2018.json
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2019.csv
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2019.json
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2020.csv
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2020.json
C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\csv\climaticos_2021.csv
C:\curso_hack_boss\Proyecto_final_AE

In [104]:

# Definimos la consulta INSERT mapeando  la tabla datos_climaticos
query = """INSERT INTO datos_climaticos ( fecha , indicativo , nombre , provincia , altitud , tmed , prec , tmin , horatmin , 
                                          tmax , horatmax , dir , velmedia , racha , horaracha , sol , presMax , horaPresMax ,
                                          presMin , horaPresMin , hrMedia , hrMax , horaHrMax , hrMin , horaHrMin )                
                                 VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                            ON CONFLICT ( indicativo , fecha ) DO NOTHING;"""

# Columnas en el orden exacto esperado por la consulta SQL
columnas = [ 'fecha' , 'indicativo' , 'nombre' , 'provincia' , 'altitud' , 'tmed' , 'prec' , 'tmin' , 'horatmin' , 'tmax' , 'horatmax' ,
             'dir', 'velmedia' , 'racha' , 'horaracha' , 'sol' , 'presMax' , 'horaPresMax' , 'presMin' , 'horaPresMin' , 'hrMedia' ,
             'hrMax' , 'horaHrMax' , 'hrMin' , 'horaHrMin' ]


# Conectamos a PostgreSQL
with psycopg.connect(DATABASE_URI) as conn:
    with conn.cursor() as cursor:
        for año in range(año_inicio, año_fin + 1):
            json_file = OUTPUT_JSON_DIR / f"climaticos_{año}.json"
            
            if not json_file.exists():
                print(f"Advertencia: El archivo {json_file.name} no existe. Saltando...")
                continue
                
            print(f"Leyendo y procesando {json_file}")
            
            # Leemos el archivo JSON
            with open(json_file, "r", encoding="utf-8") as f:
                datos_json = json.load(f)
            
            # Convertimos a DataFrame para limpiar los datos
            df_año = limpiar_y_cargar_datos_fechas_todas_estaciones(datos_json)

            df_año = df_año[columnas].replace({np.nan: None})

            # Convertimos las filas a una lista de tuplas para la inserción
            registros = list(df_año.itertuples(index=False, name=None))

           # display(registros)
            
            # Ejecutamos la inserción en lote (bulk insert)
            cursor.executemany(query, registros)
            print(f"¡Éxito! Se han procesado e insertado {len(registros)} registros para el año {año}.")


Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2016.json
¡Éxito! Se han procesado e insertado 162128 registros para el año 2016.
Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2017.json
¡Éxito! Se han procesado e insertado 300412 registros para el año 2017.
Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2018.json
¡Éxito! Se han procesado e insertado 300199 registros para el año 2018.
Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2019.json
¡Éxito! Se han procesado e insertado 302279 registros para el año 2019.
Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2020.json
¡Éxito! Se han procesado e insertado 303151 registros para el año 2020.
Leyendo y procesando C:\curso_hack_boss\Proyecto_final_AEMET\POC_Truji\sheets\json\climaticos_2021.json
¡Éxito! Se han p